### ISSUE TITLE
What is this and can we request features?

### ISSUE THREAD

**Issue by CryoKeen** [SamVanheer/HLEnhanced#2]
Hello I am very new to GitHub just signed up so I could comment on your page/send you a message 

But yeah I would like to get involved in coding and diversify my portfolio, I want to move away from the Sven Co-op engine if possible. To that end your HL:Enhanced might do that. My issue is will you create some basic svencoop stuff if we request it? Ie cool basic Sven Co-op stuff: squadmakers/custom models/custom sounds

even minimal stuff like this would help me making my own Keen Co-op type modification. Thank you and sorry if this isn't the right place for this kind of thing I couldn't find a comment section or anything.

so what are your goals for this project and what exactly is it and can I help in anyway I am somewhat very skilled at goldsource level design.
- Haeley "CryoKeen" Merrill-Schroeppel
  GoldSource Level Designer (15 years Experience) Yolo Swag Yolo

**Comment by SamVanheer** [comment]
Right now it's just code cleanup, documentation and redesign. I can add custom model and sound support easily enough, but that won't happen right away. Squadmakers will probably just be modified monstermakers. 

I want to avoid being accused of stealing code though, and adding those features could easily cause that.

**Comment by lucesenbabilonia** [comment]
It's not like Spirit of Half-Life didn't have Model Replacement already, but they might attempt to shut down any project that is backwards compatible with Sven Co-op maps.

**Comment by JoelTroch** [comment]
Furthermore, there aren't 87190787804655570570 ways to add custom models/sounds support to entities, most of the HL1 mods source codes I've seen so far used the code of Laurie Cheers (SoHL's author). I've never looked "inside" Sven-Coop, but I think they do this the SoHL way too.

@lucesenbabilonia : I highly doubt this project will be able to run SC maps, Sven-Coop has it's own flavour of the Gold Source engine so they have more "freedom" than HL: Enhanced.

**Comment by lucesenbabilonia** [comment]
@JoelTroch Pre-Sven Co-op 5 maps* :P

**Comment by JoelTroch** [comment]
@lucesenbabilonia Technically speaking, you can take a map from mod A (let's say "They Hunger") and make it run on mod B (let's say "Opposing Force") and if it's not enough, "RipENT" it. So anyone can take a SC4 map and run it under HL, Opposing Force, Blue-Shift, DMC, Ricochet or whatever he pleases.

**Comment by CryoKeen** [comment]
Hello again!

Have you already, are you working on or could you add "map will not crash if a model/sprite resource is missing" feature? :D

**Comment by SamVanheer** [comment]
That's impossible to 100% prevent when i only have access to game code. Metamod might precache assets that the game doesn't, so i can't account for that. Other than that, i can track all precaches locally and check that list before calling SetModel. Then i can fall back to ERROR if there is no precached model. It's needlessly expensive due to the fact that the engine also has that info, but that's how it can be prevented.

**Comment by CryoKeen** [comment]
Could you explain what you mean by "expensive"? It is time consuming it sounds like, sounds like it's indeed redundant but perhaps necessary without the "sourcecode". Does implementing this feature cost engine resources that could be better preserved to be used one other features/bug fixes?

Also how come Valve will not release their GoldSource code to everyone? Do they just want to stranglehold onto it? Quake 1 code came out shorter than this!

**Comment by SamVanheer** [comment]
It's expensive because you need a list of MAX_MODELS & MAX_SOUNDS strings. You'd be throwing away 133120 bytes per list, assuming they both have a limit of 512. For something that's already there, it's expensive. You'd also need to look up the resource every time it's being set or used, and that requires some tricks. 
You either need a hash map or a binary search, and the latter requires you to block precache calls that happen after MapActivate has been called. That should always be done if you want to put a stop to the late precache shutdown messages, but it adds to the complexity.

I don't know why Valve won't release it, but my guess is legal issues due to secret data encryption algorithms in the engine.

**Comment by CryoKeen** [comment]
Request: trigger_random & trigger_random_time (rename them if you like to avoid IP)

Also me and my friends were having trouble because of the name of the mod. I believe it was mixed uppercase and lowercase, so we ended up renaming it and had miss-matched version errors, maybe you could all lowercase the folder name?

Thanks your doing great work!

**Comment by CryoKeen** [comment]
New Request for HL:EH and Prototype Engine: Func_Vehicle

**Comment by tenyuhuang** [comment]
Hi Solokiller, nice to see you've been working on a HL fork. Any chance of getting a full Unicode support? Just asking. Keep up the great job!

**Comment by JoelTroch** [comment]
@CryoKeen Can you detail how these trigger_random work ? func_vehicle like CS 1.6 would be possible.

@tenyuhang Unless @SamVanheer tells me I'm wrong, this would require engine access.

**Comment by SamVanheer** [comment]
trigger_random selects a random target from a group of targets, there are variants for time based triggering and unique triggering (trigger every target once before the same target can be triggered again).

Once custom entities are implemented the script can be modified from SC's version to one that works with HLE.

Unicode support depends on the engine, VGUI1 doesn't support Unicode at all, but VGUI2 does to a certain extent. We can't do much about it now.

**Comment by SamVanheer** [comment]
I've implemented trigger_random, latest Windows binaries have been committed, the entity has also been added to the fgd. Its behavior is identical to Sven Co-op's version, it uses the spawnflag based time and unique behavior, no trigger_random_time is needed. Their entity documentation should tell you how it works, the spawn flags have different values though, so be aware of that.

No Sven Co-op programmer's feelings were hurt in the making of this entity.
